### ISSUE TITLE
validate the completeness of the potential-targets vs observed-targets tables

### ISSUE THREAD

**Issue by moustakas** [moustakas/desi-photometry#3]
@weaverba137 @sbailey

Need to double-check that:

1. If every target in targetphot-edr.fits is also in targetphot-potential-edr.fits AND
2. Every row in targetphot-edr.fits has identical values to the corresponding row in targetphot-potential-edr.fits, THEREFORE
3. One could create a 100% comprehensive target photometry table in a database by ONLY ingesting  targetphot-potential-edr.fits.

**Comment by moustakas** [comment]
I ran the following little script and found some differences in the sv1, sv2, and sv3 surveys that indicate an issue (sigh). 

In sv1 there are differences in SUBPRIORITY, OBSCONDITIONS, PRIORITY_INIT, NUMOBS_INIT, SV1_DESI_TARGET, SV1_MWS_TARGET, and SV1_SCND_TARGET. And in sv2 and sv3 there are differences in SUBPRIORITY, PRIORITY_INIT, and NUMOBS_INIT. So I need to chase these issues down before we move the catalogs into their final location. @weaverba137 @sbailey @araichoor @geordie666 @dstndstn

```python
for survey in ['cmx', 'special', 'sv1', 'sv2', 'sv3']:
    print('Checking survey {}'.format(survey))
    allgood = True
    obs = Table(fitsio.read('observed-targets/targetphot-{}-fuji.fits'.format(survey)))
    pot = Table(fitsio.read('potential-targets/targetphot-potential-{}-fuji.fits'.format(survey)))
    I = np.isin(pot['TARGETID'], obs['TARGETID'])
    for col in pot.colnames:
        diff = pot[I][col] != obs[col]
        if np.any(diff):
            J = ~np.isnan(obs[col][diff]) # need to check nan separately
            if np.sum(J) > 0:
                allgood = False
                print(col, pot[I][col][diff][J], obs[col][diff][J])
    if allgood:
        print('  No differences found.')
```

**Comment by weaverba137** [comment]
Thank you for running these checks. A few points:

1. For the sake of comprehensive quality assurance you should also check the summary files, not just the per-survey files.
2. You should check that in each *individual* file that `TARGETID` is unique, *e.g*: `assert len(np.unique(obs['TARGETID'])) == len(obs['TARGETID'])`. Once you've done that, you can add `assume_unique=True` to `np.isin()` to speed things up.
3. You could easily incorporate the check "Is every target in the `targetphot` file also in the `targetphot-potential` file", but I don't see that here. You should do something like `K = np.isin(obs['TARGETID'], pot['TARGETID'], assume_unique=True); assert len(K) = len(obs['TARGETID'])`.

**Comment by moustakas** [comment]
TARGETID can be observed across surveys *as a different type of target*. E.g., the same target can be primary in one survey and secondary in another survey, in which case its `targetphot` table will be populated differently. Hence, `survey` has to be included in the matching, which is also the cause of much of the heartache of putting these catalogs together.

**Comment by weaverba137** [comment]
That's a good clue, but a quick test I did yesterday on `targetphot-potential-edr.fits` seemed to indicate that `TARGETID` was unique. Is that possibly a result of the construction of that file rather than the actual uniqueness or non-uniqueness of `TARGETID`?

**Comment by araichoor** [comment]
if useful, quick thoughts/remarks (without having looked at the data):
- SUBPRIORITY: I don t how s that handled in desitarget across surveys; however, remember that until the early main, the SUBPRIORITY was accidently re-written at the fiberassign step; so it would unfortunately make sense to have different values if a given TARGETID appears in different tiles of the same survey;
- SV1_DESI_TARGET, SV1_MWS_TARGET, and SV1_SCND_TARGET: here, I m a bit more surprised; would you have few examples?

**Comment by moustakas** [comment]
I think I've found the issue in my code (which only impacts potential targets the way my code is structured, I think). Within a given survey, I was assuming that TARGETID was unique, but of course it's not. Here's an example of the same target which got assigned to two different tiles/programs:
```
<Table length=2>
     TARGETID     TILEID SURVEY PROGRAM
      int64       int32   str3    str6
----------------- ------ ------ -------
39633236187415756  81000    sv2    dark
39633236187415756  81056    sv2  bright
```
So the fiberassign quantities are different:
```
<Table length=2>
     TARGETID         SUBPRIORITY     PRIORITY_INIT NUMOBS_INIT
      int64             float64           int64        int64
----------------- ------------------- ------------- -----------
39633236187415756 0.29006353007782204          2000           1
39633236187415756   0.844270126744655          3200           1
```
I'm going to rerun now but this check demonstrates that I need to include SURVEY, PROGRAM, and TILEID in the output `potential-targets table`.

**Comment by araichoor** [comment]
I confirm that `TARGETID= 39633236187415756` passes the following sv2 selections: `LRG and BGS_FAINT`.
So it makes sense that it inherits the LRG desitarget settings (`PRIORITY_INIT,  NUMOBS_INIT`) for the dark observations (which overrule the BGS_FAINT settings), but it gets the BGS_FAINT desitarget settings for the bright observations.
Nice catch.

For the `SUBPRIORITY`: the values actually are different for the dark/bright desitarget catalogs:
```
>>> from desitarget.io import read_targets_in_tiles
>>> tid = 39633236187415756
>>> tiles = Table.read("/global/cfs/cdirs/desi/survey/fiberassign/SV2/20210324/081000-tiles.fits")
```
Reading the sv2 dark desitarget catalogs:
```
>>> d["SV2_DESI_TARGET", "SV2_BGS_TARGET", "OBSCONDITIONS", "PRIORITY_INIT", "NUMOBS_INIT", "SUBPRIORITY"][d["TARGETID"] == tid]
<Table length=1>
  SV2_DESI_TARGET   SV2_BGS_TARGET OBSCONDITIONS PRIORITY_INIT NUMOBS_INIT    SUBPRIORITY   
       int64            int64          int64         int64        int64         float64     
------------------- -------------- ------------- ------------- ----------- -----------------
1152921504606847233            257           519          3200           1 0.844270126744655
```
Reading the sv2 bright desitarget catalogs:
```
>>> d = Table(read_targets_in_tiles("/global/cfs/cdirs/desi/target/catalogs/dr9/0.53.0/targets/sv2/resolve/bright", tiles=tiles, quick=True))
>>> d["SV2_DESI_TARGET", "SV2_BGS_TARGET", "OBSCONDITIONS", "PRIORITY_INIT", "NUMOBS_INIT", "SUBPRIORITY"][d["TARGETID"] == tid]
<Table length=1>
  SV2_DESI_TARGET   SV2_BGS_TARGET OBSCONDITIONS PRIORITY_INIT NUMOBS_INIT     SUBPRIORITY    
       int64            int64          int64         int64        int64          float64      
------------------- -------------- ------------- ------------- ----------- -------------------
1152921504606847233            257           519          2000           1 0.29006353007782204
```

**Comment by sbailey** [comment]
I was initially surprised by having the same target appear under two different priorities, but after reloading some memories from backup I'm pretty sure this was purposeful design to keep the bright and dark surveys isolated from each other.  e.g. so that BGS targets that happen to pass LRG cuts don't end up becoming an extra-high priority special class of targets within the bright time program.

**Comment by moustakas** [comment]
This validation process was very useful as it uncovered a handful of additional issues that I've now fixed. 

For completeness (and to spare ourselves this effort in the future), I've added a [validation script](https://github.com/moustakas/desi-photometry/blob/main/validate-files) to the repo which collects the statistics (numbers of objects, file sizes, etc.) which are added to the README--
https://github.com/moustakas/desi-photometry/blob/main/README.md

In addition, the validation script ensures that *for a given SURVEY and TILEID*:
* The set of observed and potential TARGETIDs are unique;
* Every object in the `observed-targets/targetphot-SURVEY-SPECPROD.fits` catalog exists (and has one and only one entry) in the `potential-targets/targetphot-potential-SURVEY-SPECPROD.fits` file; and
* The data recorded for the overlapping potential and observed targets are identical.

**Important**: Since the same TARGETID can appear across surveys *and/or* across TILEIDs (within a given SURVEY), the following quantities need to be specified when seeking out the data for a given object: `SURVEY`, `TILEID`, and `TARGETID`, which are now all present in all the `targetphot` tables (but were not previously). (For convenience, I've also added `PROGRAM` to these files.)

@sbailey, unless you, @weaverba137, or others want some time to check my work (which is always welcome!), the VACs are ready to be moved to their final location:

```
/global/cscratch1/sd/ioannis/lsdr9-photometry/fuji --> /global/cfs/cdirs/desi/public/edr/vac/lsdr9-photometry/
/global/cscratch1/sd/ioannis/lsdr9-photometry/guadalupe --> /global/cfs/cdirs/desi/public/dr1/vac/lsdr9-photometry/
```

Once the files have been moved into place I'll need to update the tutorial notebook
https://github.com/moustakas/desi-photometry/blob/main/example.ipynb
and then I'll retag this repo `fujilupe-v1.0`.
